testPassingInvalidSizeToSetComplexSocketOptionFails flaky

weird 'out of memory' failure in virtualised macOS

```
Test Case '-[NIOTests.SocketOptionProviderTest testPassingInvalidSizeToSetComplexSocketOptionFails]' started.
/private/tmp/task-sandbox/-7497412305172451960.V4H4jt/0~vGyWzLVgFkF3b1-I_8usZW9BljRgcBsESxOp0H8x74_B_5DQUoNVy0bnPnok2DJumZvVn60yZSXgJUzcjTpT9w==/SRCROOT/Tests/NIOTests/SocketOptionProviderTest.swift:64: error: -[NIOTests.SocketOptionProviderTest testPassingInvalidSizeToSetComplexSocketOptionFails] : failed - unexpected error setsockopt(socket:level:optionName:optionValue:optionLen:) failed: Cannot allocate memory (errno: 12) thrown
Fatal error: Unexpectedly found nil while unwrapping an Optional value
```

CC'ing @Lukasa

--- a/Tests/NIOTests/SocketOptionProviderTest.swift
+++ b/Tests/NIOTests/SocketOptionProviderTest.swift
@@ -58,15 +58,25 @@ final class SocketOptionProviderTest: XCTestCase {
         self.serverChannel = try? assertNoThrowWithValue(ServerBootstrap(group: group).bind(host: "127.0.0.1", port: 0).wait())
         self.clientChannel = try? assertNoThrowWithValue(ClientBootstrap(group: group).connect(to: serverChannel.localAddress!).wait())
 
+        // We need to join these multicast groups on the loopback interface to work around issues with rapidly joining and leaving
+        // many multicast groups. On some OSes, if we do that on a public interface, we can build up a kernel backlog of IGMP
+        // joins/leaves that may eventually lead to an ENOMEM and a spurious test failure. As joining/leaving groups on loopback
+        // interfaces does not require IGMP joins/leaves, forcing these joins onto the loopback interface saves us from this
+        // risk.
+        let v4LoopbackAddress = try! SocketAddress(ipAddress: "127.0.0.1", port: 0)
+        let v6LoopbackAddress = try! SocketAddress(ipAddress: "::1", port: 0)
+        let v4LoopbackInterface = try! System.enumerateInterfaces().filter { $0.address == v4LoopbackAddress }.first!
+
         self.ipv4DatagramChannel = try? assertNoThrowWithValue(
             DatagramBootstrap(group: group).bind(host: "127.0.0.1", port: 0).then { channel in
-                return (channel as! MulticastChannel).joinGroup(try! SocketAddress(ipAddress: "224.0.2.66", port: 0)).map { channel }
+                return (channel as! MulticastChannel).joinGroup(try! SocketAddress(ipAddress: "224.0.2.66", port: 0), interface: v4LoopbackInterface).map { channel }
             }.wait()
         )
 
         // The IPv6 setup is allowed to fail, some hosts don't have IPv6.
+       let v6LoopbackInterface = try! System.enumerateInterfaces().filter { $0.address == v6LoopbackAddress }.first
         self.ipv6DatagramChannel = try? DatagramBootstrap(group: group).bind(host: "::1", port: 0).then { channel in
-            return (channel as! MulticastChannel).joinGroup(try! SocketAddress(ipAddress: "ff12::beeb", port: 0)).map { channel }
+            return (channel as! MulticastChannel).joinGroup(try! SocketAddress(ipAddress: "ff12::beeb", port: 0), interface: v6LoopbackInterface).map { channel }
         }.wait()
     }
 

